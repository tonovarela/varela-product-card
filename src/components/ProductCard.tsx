import React from 'react';
import styles from '../styles/styles.module.css';

import { useProduct } from './../hooks/useProduct';
import { InitialValues, OnChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/interfaces';
import { createContext } from 'react';


export const ProductContext = createContext({} as ProductContextProps)
export const { Provider } = ProductContext;

export interface Props {
    product: Product;
    className?: string;    
    children:(args:ProductCardHandlers)=>JSX.Element ;
    style?:React.CSSProperties;
    onChange?: (args:OnChangeArgs) => void ;
    value?:number;
    initialValues?:InitialValues
}


export const ProductCard = ({ children, product, className ,style={},onChange,value=0,initialValues}: Props) => {
    const { counter, increaseBy,isMaxReached,reset } = useProduct( { onChange,product,value,initialValues});

    return (

        <Provider value={{ counter, increaseBy, product ,isMaxReached}}>
            <div className={`${styles.productCard} ${className}`}  style={style}>
                {children({
                    count:counter,
                    isMaxCountReached:isMaxReached(),
                    maxCount:initialValues?.maxCount,
                    increaseBy,
                    product,
                    reset
                })}
            </div>
        </Provider>

    )
}

