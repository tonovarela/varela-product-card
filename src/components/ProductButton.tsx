import React from "react";
import {  useContext } from "react";

import styles from '../styles/styles.module.css';
import { ProductContext } from "./ProductCard";

export interface Props {
    className?: string;
    style?: React.CSSProperties
}

export const ProductButtons = ({ className = '', style }: Props) => {
    const { increaseBy, counter, isMaxReached} = useContext(ProductContext);
    //const isMaxCountReached = useCallback(() => !!maxCount && counter === maxCount, [counter, maxCount])
    const styleIncreaseButton = `${styles.buttonAdd} ${isMaxReached() ? styles.disabled : ''} `;
    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button className={`${styles.buttonMinus}`} onClick={() => increaseBy(-1)}> - </button>
            <div className={styles.countLabel}>{counter}</div>
            <button className={styleIncreaseButton} disabled={isMaxReached()} onClick={() => increaseBy(+1)}> + </button>
        </div>
    )
}