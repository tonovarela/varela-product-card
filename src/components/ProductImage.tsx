import React from "react";
import { useContext } from "react"


import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import { ProductContext } from "./ProductCard";

export interface Props {
    img?:string;
    className?:string;
    style?:React.CSSProperties
}
export const ProductImage = ({ img = '',className ='',style }:Props) => {
    const { product } = useContext(ProductContext)
    const imagen =img.length > 0?img :product.img    
    return (
        <img className={ `${styles.productImg} ${className} `}  style={style}  src={imagen ?imagen : noImage} alt="Product"></img>
    )
}