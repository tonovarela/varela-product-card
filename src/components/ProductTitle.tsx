import React from "react";
import { useContext } from "react";

import styles from '../styles/styles.module.css';
import { ProductContext } from "./ProductCard";

export interface Props {
    title?:string;
    className?:string;
    style?:React.CSSProperties;
}
export const ProductTitle = ({ title = '',className='' ,style}:Props) => {
    const { product } = useContext(ProductContext)
    let titulo=title.length > 0?title:product.title;    
    return (
        <span className={`${styles.productDescription} ${className}`} style={style}>{titulo}</span>
    )
}