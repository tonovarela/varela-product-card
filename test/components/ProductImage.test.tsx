import React from "react";
import renderer from "react-test-renderer";

import {  ProductCard,ProductImage } from '../../src/components'
import {  product2 } from './../data/product';

describe('ProductImage', () => {
    test('Se debe de mostrar el componente del la imagen producto', () => {        
        const wrapper = renderer.create(
            <ProductCard product={product2} >
                {
                    () => (<ProductImage  />)
                }
            </ProductCard>
        );
        expect(wrapper.toJSON()).toMatchSnapshot()
    })

    test('Se debe de mostrar el componente del la imagen personalizada', () => {        
        const wrapper = renderer.create(
            <ProductCard product={product2} >
                {
                    () => (<ProductImage  img="http://logo.com" />)
                }
            </ProductCard>
        );
        expect(wrapper.toJSON()).toMatchSnapshot()
    })
});