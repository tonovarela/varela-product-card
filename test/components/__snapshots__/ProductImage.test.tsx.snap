// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Se debe de mostrar el componente del la imagen personalizada 1`] = `
<div
  className="productCard undefined"
  style={Object {}}
>
  <img
    alt="Product"
    className="productImg  "
    src="http://logo.com"
  />
</div>
`;

exports[`ProductImage Se debe de mostrar el componente del la imagen producto 1`] = `
<div
  className="productCard undefined"
  style={Object {}}
>
  <img
    alt="Product"
    className="productImg  "
    src="coffee-mug.png"
  />
</div>
`;
